sdk/search/search/review/search.api.md

```diff
@@ -246,11 +817,113 @@ export type SuggestResult<T> = {
     readonly text: string;
 } & T;
 
+// @public
+export interface SynonymTokenFilter {
+    expand?: boolean;
+    ignoreCase?: boolean;
+    name: string;
+    odatatype: "#Microsoft.Azure.Search.SynonymTokenFilter";
+    synonyms: string[];
+}
+
+// @public
+export interface TagScoringFunction {
+    boost: number;
+    fieldName: string;
+    interpolation?: ScoringFunctionInterpolation;
+    parameters: TagScoringParameters;
+    type: "tag";
+}
+
+// @public
+export interface TagScoringParameters {
+    tagsParameter: string;
+}
+
+// @public
+export interface TextWeights {
+    weights: {
+        [propertyName: string]: number;
+    };
+}
+
+// @public
+export type TokenCharacterKind = 'letter' | 'digit' | 'whitespace' | 'punctuation' | 'symbol';
+
+// @public
+export interface TokenFilter {
+    name: string;
+    odatatype: "TokenFilter";
+}
+
+// @public
+export type TokenFilterName = 'arabic_normalization' | 'apostrophe' | 'asciifolding' | 'cjk_bigram' | 'cjk_width' | 'classic' | 'common_grams' | 'edgeNGram_v2' | 'elision' | 'german_normalization' | 'hindi_normalization' | 'indic_normalization' | 'keyword_repeat' | 'kstem' | 'length' | 'limit' | 'lowercase' | 'nGram_v2' | 'persian_normalization' | 'phonetic' | 'porter_stem' | 'reverse' | 'scandinavian_normalization' | 'scandinavian_folding' | 'shingle' | 'snowball' | 'sorani_normalization' | 'stemmer' | 'stopwords' | 'trim' | 'truncate' | 'unique' | 'uppercase' | 'word_delimiter';
+
+// @public
+export type TokenFilterUnion = TokenFilter | AsciiFoldingTokenFilter | CjkBigramTokenFilter | CommonGramTokenFilter | DictionaryDecompounderTokenFilter | EdgeNGramTokenFilter | EdgeNGramTokenFilterV2 | ElisionTokenFilter | KeepTokenFilter | KeywordMarkerTokenFilter | LengthTokenFilter | LimitTokenFilter | NGramTokenFilter | NGramTokenFilterV2 | PatternCaptureTokenFilter | PatternReplaceTokenFilter | PhoneticTokenFilter | ShingleTokenFilter | SnowballTokenFilter | StemmerTokenFilter | StemmerOverrideTokenFilter | StopwordsTokenFilter | SynonymTokenFilter | TruncateTokenFilter | UniqueTokenFilter | WordDelimiterTokenFilter;
+
+// @public
+export interface TokenInfo {
+    readonly endOffset: number;
+    readonly position: number;
+    readonly startOffset: number;
+    readonly token: string;
+}
+
+// @public
+export interface Tokenizer {
+    name: string;
+    odatatype: "Tokenizer";
+}
+
+// @public
+export type TokenizerName = 'classic' | 'edgeNGram' | 'keyword_v2' | 'letter' | 'lowercase' | 'microsoft_language_tokenizer' | 'microsoft_language_stemming_tokenizer' | 'nGram' | 'path_hierarchy_v2' | 'pattern' | 'standard_v2' | 'uax_url_email' | 'whitespace';
+
+// @public
+export type TokenizerUnion = Tokenizer | ClassicTokenizer | EdgeNGramTokenizer | KeywordTokenizer | KeywordTokenizerV2 | MicrosoftLanguageTokenizer | MicrosoftLanguageStemmingTokenizer | NGramTokenizer | PathHierarchyTokenizerV2 | PatternTokenizer | StandardTokenizer | StandardTokenizerV2 | UaxUrlEmailTokenizer;
```

Comment:
I believe that no, we don't consider it a breaking change, though it would probably be a minor version bump since it might cause a TS build break if their code can't narrow something anymore.

Reply:
@bterlson for thoughts

Reply:
Pedantically, adding new cases to a union is a breaking change. For example:

```ts
export type TokenizerUnion = { a: 1, b: string } | { a: 2, b: string };
// build break if you add  {a: 1, b: number}

function handleTokenizer(t: TokenizerUnion): number {
    if (t.a === 1) {
        return t.b.length;
        // but also in JS land, this will throw if t is {a: 1, b: number}
    }

    return -1;
}
```

This is the type system warning you about a JS-level and service-level breaking change - the code above will not work properly with the new member of the union.

That said, we can't be cutting new majors any time we add a new constituent to a discriminated union. I think we need to consider such additions on a case-by-case basis. My guess is most of the time it will not be a concern, as code users write will in practice not look like the above.

Reply:
@bterlson In our API, these models typically represent parts of resources that are managed via CRUD operations. New cases will be added as we add new features to the API, and the typical usage would be to specify new cases as input to a Create/Update operation. While it's possible someone might write code to dispatch over the different cases returned from a Get or List operation, it seems unlikely.

In general, I'd like to see the versioning and backward compatibility guidelines clarified across all languages. They should specify the different types of breakage for each language (e.g. -- in .NET, we have binary, source, and behavioral compatibility), which warrant major/minor version bumps, and which we should avoid altogether. As it stands, the versioning guidelines alternate between vague and super-conservative to the point where adding features to the service in a compliant way would be a challenge.

Reply:
Definitely agree the guidelines should be clarified (I've added a stub section to our guidelines to fill out later with more of these details), but it is also impossible in my experience to, for example, create a flow chart you can follow to know what the best course of action is. It will always involve some grey matter I think.